apps/sm2_threshold.c

```diff
@@ -0,0 +1,434 @@
+/*
+ * Copyright 2023 The Tongsuo Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://github.com/Tongsuo-Project/Tongsuo/blob/master/LICENSE.txt
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include "apps.h"
+#include "progs.h"
+#include <openssl/bio.h>
+#include <openssl/err.h>
+#include <openssl/pem.h>
+#include <openssl/rand.h>
+#include <openssl/sm2_threshold.h>
+#include <openssl/ec.h>
+#include <internal/cryptlib.h>
+
+#undef BUFSIZE
+#define BUFSIZE 1024*8
+
+typedef enum OPTION_choice {
+    OPT_COMMON,
+    OPT_DERIVE, OPT_KEY, OPT_KEY1, OPT_PUBKEY2, OPT_PUBOUT, OPT_SIGN1,
+    OPT_SIGN2, OPT_SIGN3, OPT_TEMP_PUBKEY, OPT_DIGEST, OPT_TEMP_KEY, OPT_OUT,
+    OPT_PARTIAL_SIGNATURE, OPT_PASSIN,
+} OPTION_CHOICE;
+
+/*
+Examples:
+sm2_threshold -derive -key key1 -pubout pubkey1
+sm2_threshold -derive -key1 key1 -pubkey2 pubkey2 -pubout pubkey
+sm2_threshold -sign1 pubkey1 file
+sm2_threshold -sign2 key2 -digest dgst -temp_pubkey pub -out out
+sm2_threshold -sign3 key1 -partial_signature partial_sig -temp_key key -out out
+*/
```

Comment:
需要改成能体现两方的例子，比如：

~~~
Assume there are two sides A and B performing SM2 threshold signature.

On A:
sm2_threshold -derive -key private_a -pubout pub_a
...send pub_a to B, and get pub_b from B...
sm2_threshold -derive -key1 private_a -pubkey2 pub_b -pubout pub_whole

On B:
sm2_threshold -derive -key private_b -pubout pub_b
...send pub_b to A, and get pub_a from A...
sm2_threshold -derive -key1 private_b -pubkey2 pub_a -pubout pub_whole

Thus, A and B all have the same complete public key 'pub_whole', it will be used in SM2 verification.

Then A and B can produce a signature on a given message 'file':

On A:
sm2_threshold -sign1 pub_a file   -->  will have what? a file named digest???
...then send output to B...
....
~~~

Reply:
Fixed